README.md
<!-- https://github.com/seomoz/docs/blob/improved-readme/Readme%20header.md -->

![Status Incubating][status-badge]
![Team Big Data][team-badge]
![Closed Source][open-badge]
![Not Critical][critical-badge]
[![Email][email-badge]][email-link]
[![Slack][slack-badge]][slack-link]
![Internal][scope-badge]

[status-badge]: https://img.shields.io/badge/status-incubating-blue.svg?style=flat
[team-badge]: https://img.shields.io/badge/team-big_data-green.svg?style=flat
[open-badge]: https://img.shields.io/badge/open_source-nope-orange.svg?style=flat
[critical-badge]: https://img.shields.io/badge/critical-no-lightgrey.svg?style=flat
[email-badge]: https://img.shields.io/badge/email-bigdata--dev%40moz.com-green.svg?style=flat
[email-link]: mailto:bigdata-dev@moz.com
[slack-badge]: https://img.shields.io/badge/slack-%23big--data-ff69b4.svg?style=flat
[slack-link]: https://moz.slack.com/messages/big-data/
[scope-badge]: https://img.shields.io/badge/scope-internal-lightgrey.svg?style=flat

# Fiji

This workspace contains the collection of Fiji projects.

## Prerequisites

Make sure you have the following installed:

 - Python 3.4.2 or greater.
 - Oracle JDK 7 or greater.

## Build instructions

 1. Download and extract the workspace into a directory `$wkspc`:  
    ```git clone https://github.com/seomoz/fiji $wkspc```

 2. Make sure the workspace settings are correct in:
    [`$wkspc/.workspace_config/conf.py`](https://github.com/seomoz/fiji/tree/master/.workspace_config/conf.py)

 3. From the workspace root directory, the Fiji artifacts (Maven artifacts and CLI executables) can be build by running:
    ```./bin/fiji-build build```

    The build process fetches and produces Maven artifacts into the local repository: `$wkspc/output/maven_repository/`.
    Executables are generated into `$wkspc/output/bin/`.


## Fiji CLI

Upon successful build, the Fiji CLI may be found at `$wkspc/output/bin/com/moz/fiji/schema/fiji`.
Different versions of the CLI may be found at `$wkspc/output/bin/com/moz/fiji/schema/fiji-cdh5.x`.

In order to (re)build a specific version of the CLI, you may use:
```./bin/fiji-build build //com/moz/fiji/schema:fiji-cdh5.x```.

The definition of the Fiji artifacts can be found in the BUILD descriptor in the root directory: [`$wkspc/BUILD`](https://github.com/com/moz/fiji/tree/master/BUILD).


## Running tests

In order to run tests, you can use:
```./bin/fiji-build test```

Tests involving the Hadoop Map/Reduce framework may run very slowly unless the appropriate version of the Hadoop native libraries are installed on the system.


fiji-rest/RELEASE_NOTES.txt

                              FijiREST Release Notes

Version 1.5.0
  * REST-144. Split the standard plugin into its own module to allow for the
      "/v1" endpoint to be removed/disabled.
  * POM-40. Require JDK 7 to compile Fiji projects.
  * REST-135. Support bulk gets
  * REST-147. Fix the startup script to make the logs directory and
       ensure logs are written there.

Version 1.4.0
  * REST-133. Make Fiji/FijiTable connection expiration configurable
  * REST-105. Add endpoint to gracefully (or forcefully) shut down server.
  * REST-137. Fiji-REST now supports DELETE commands
  * REST-98.  Re-introduce control on Fiji instance visibility in FijiREST

Version 1.2.2
  * REST-90. Allow FijiREST to be started in foreground mode, via the
      'foreground' command line option.

Version 1.2.0
  * REST-62. Implements counter cell incrementation and decrementation.
  * REST-41. Supports optional cross-origin resource sharing (CORS).
  * REST-89. Refresh instances automatically. Removes instance visibility
      configuration parameter. All instances on target cluster are visible.
      Manual refreshes can be triggered with a POST to the admin task
      /tasks/refresh_instances.

Version 1.1.0
  * REST-85. Add archetype for creating Fiji Rest plugins.
  * REST-63. FijiREST should have a mode to serve all Fiji instances in a cluster.
  * REST-84. fiji-rest script should allow specifying JVM arguments.
  * REST-76. FijiREST logging locations must be fully configurable
  * REST-79. Don't close the underlying Fiji instance.
  * REST-80. Correctly use FIJI_REST_CONF_DIR for location of conf directory.
  * REST-69. Modularizes FijiREST in preparation for custom endpoints.

Version 1.0.1
  * REST-50. Fix a performance issue that was causing massive slowdowns.

Version 1.0.0
  * REST-66. Incorporates native json formatted entity ids in POST.
      Cleans up row key specification language.
  * REST-64. Fix the exception handling to ensure that all exceptions
      are converted to JSON objects, stacktraces included and logged
      to the application log file on the server.
  * REST-48. Adds start_eid/end_eid and removes start_rk/end_rk.
  * REST-60. Removes the entityId endpoint of tables.
  * REST-50. Check that the table still exists before returning it in
      ManagedFijiClient.
  * REST-59. Ensure that if the writer_schema is JSON, that the server will
     accept the schema and parse it properly.

Version 0.6.0
  * REST-55. Remove an old servlet dependency. Also change the release artifact
     to not contain a single fat jar but house the runtime dependencies in the
     lib/ directory.
  * REST-54. Supports per-request freshening parameters.
  * REST-53. Remove the row endpoint. Can be accessed by /rows?eid=.
  * REST-32. Supports scanning with partially specified entity ids.
  * REST-49. Adds freshening/scoring parameter to FijiREST.

Version 0.5.0
  * REST-44. Remove the rowkey element from the output of the row.
  * REST-45. Output the writer schema_id on GET. Force the specification of the
      writer schema (either id or JSON) on POST to ensure that data specified can
      be written without throwing exceptions because of data being written
      according to the wrong writer schema.

Version 0.4.0
  * REST-40. Bugfix: Fix the issue where /instances/<instance> threw a
      500 error. This was due to an issue in converting Avro to JSON using
      the generic writer. Specific writers are now used when serializing a
      specific record.

Version 0.3.0
  * REST-21. Implements batch POST.
  * REST-33. Bugfix: instances/<instance>/ throws an error retrieving metadata.
  * REST-23. Expose methods to read and write REST row models.
  * REST-12. Emit JSON that is properly navigable instead of a list of cells.
  * REST-8. Allow for the installation of custom HBase configuration via either
      $BASEDIR/conf or the FIJI_REST_CONF_DIR environment variable.

Version 0.2.0
  * REST-20. Update FijiREST to depend on FijiSchema v1.1.0.
  * REST-10. Ensure that the /rows resource streams properly in all cases.
  * REST-18. Adds FIJI_CLASSPATH to FijiREST classpath.
  * REST-9. Allows users to GET with timerange=all.
  * REST-11. Ignore cells with negative timestamps when POSTing.

Version 0.1.0
  * Initial code commit.


fiji-rest/clients/ruby/lib/fijirest/client.rb
# (c) Copyright 2013 WibiData, Inc.
#
# See the NOTICE file distributed with this work for additional
# information regarding copyright ownership.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require "fijirest/version"
require 'net/http'
require 'uri'
require 'json'
require 'cgi'

module FijiRest

  # Provides a simple wrapper interface around the FijiREST server. The return of most
  # of the methods are parsed JSON.
  class Client

    # Some global helpful constants for clients.
    ENTITY_ID = "entityId"

    # Error class that wraps exceptions thrown by the server
    class FijiRestClientError < StandardError
      attr_reader :json_error_message
      def initialize(json_error)
        @json_error_message =  json_error
      end

      def to_s
        json_error_message.to_json
      end
    end

    #
    # Constructor for the FijiRest::Client. The only optional argument is the base_uri
    # defining the location of the server.
    def initialize(base_uri= "http://localhost:8080")
      @base_uri = base_uri
      @version = "v1"
    end

    #
    # Returns a list of visible instances
    def instances
      get_json(instances_endpoint)
    end

    #
    # Returns the instance description
    # param: instance_name is the name of the instance to describe.
    def instance(instance_name)
      get_json(instance_endpoint(instance_name))
    end

    # Returns a single row (identified by the specified entity_id) from the given table.
    # param: instance_name is the name of the instance
    # param: table_name is the name of the table
    # param: entity_id is an array of components comprising the entity_id to fetch.
    # param: filters is a hash containing any filters to apply which will get translated as
    #        query parameters in the REST call.
    def row(instance_name, table_name, entity_id, filters= {})
      eid_filter = if(entity_id.is_a? Array)
        entity_id.to_json
      else
        entity_id
      end
      get_json(rows_endpoint(instance_name, table_name), filters.merge({:eid => eid_filter}))
    end

    # Returns a list of the tables in the given instance.
    # param: instance_name is the name of the instance
    def tables(instance_name)
      get_json(tables_endpoint(instance_name))
    end

    # Returns the table layout for a given table.
    # param: instance_name is the name of the instance
    # param: table_name is the name of the table
    def table(instance_name, table_name)
      get_json(table_endpoint(instance_name, table_name))
    end

    # Performs a scan (via HTTP streaming) yielding each row to the given block.
    # param: instance_name is the name of the instance
    # param: table_name is the name of the table
    # param: filters is a hash containing any filters to apply which will get translated as
    #        query parameters in the REST call.
    # param: block is an anonymous block of code that will be given a hash representing the
    #        current row.
    def rows(instance_name, table_name, filters= {}, &block)
      if !block
        raise "No block given!"
      end

      url_query_params =  filters.map {|k, v| "#{k}=#{CGI.escape(v.to_s)}"}.join("&")
      uri =  URI(@base_uri)
      http =  Net::HTTP.new(uri.host, uri.port)
      http.request_get("#{rows_endpoint(instance_name, table_name)}?#{url_query_params}") \
       do |response|
        case response
        when Net::HTTPSuccess then
          remainder_json_line =  ""
          response.read_body { |chunk|
            if chunk.size > 0
              # Few possible situations: Not sure what can actually happen let's prepare for all
              # 1) chunk is a line that ends with \r\n
              # 2) chunk could be multiple lines and also ends with \r\n
              # 3) chunk has multiples lines but an incomplete last line.
              remainder_json_line =  remainder_json_line + chunk
              if remainder_json_line[-2..-1] ==  "\r\n"
                json_lines =  remainder_json_line.split("\r\n")
                json_lines.each {|line|
                  yield JSON.parse(line)
                }
                remainder_json_line =  ""
              else
                json_lines =  remainder_json_line.split("\r\n")
                json_lines.slice(0..-2).each {|line|
                  yield JSON.parse(line)
                }
                remainder_json_line =  json_lines.last
              end
            end
          }
        else
          raise_exception(response.body)
        end
      end
    end

    # Creates or updates a new row by POSTing the row_hash to the server. Will create/update
    # the row specified by the entityId key in the row_hash.
    # param: instance_name is the instance
    # param: table_name is the table
    # param: row_hash is a hash containing the data to POST. Must be of the format:
    #        {
    #            "entityId" : [//JSON component Array],
    #            "cells" : {
    #               "family" : {
    #                   "qualifier" : [
    #                                     {
    #                                       "timestamp" : <timestamp>,
    #                                       "value": <JSON representation of object to post>,
    #                                       "writer_schema": <JSON representation of the writer
    #                                                         schema (Avro) OR the Fiji ID of the
    #                                                         schema>
    #                                     },
    #                                       ...
    #                                   ]
    #                }
    #              }
    #         }
    # param: strip_timestamp will remove any timestamp fields from the cells allowing the server
    #        to set the timestamp automatically (default false).
    # param: timeout_seconds specifies the number of seconds before a Timeout::Error is thrown
    #        (default nil, which falls back to Net:HTTP default of 60 seconds)
    def write_row(instance_name, table_name, row_hash, strip_timestamp= false, timeout_seconds = nil)
      if strip_timestamp
        local_row_hash = row_hash.clone
        if local_row_hash.include?("cells")
          local_row_hash["cells"].each { |family,qualifierMap|
            if qualifierMap
              qualifierMap.each_value { |cells|
                  cells.each {|cell|
                  # TODO: Remove condition when there's a better way to represent cell level error
                  # conditions on GETs.
                  cell.delete("timestamp") unless cell["timestamp"] ==  -1
                  }
                }
            end
            }
        end
        row_json =  local_row_hash.to_json
      else
        row_json =  row_hash.to_json
      end
      uri= URI(@base_uri)
      http =  Net::HTTP.new(uri.host, uri.port)
      if timeout_seconds
        http.read_timeout = timeout_seconds
      end
      response =  http.post(rows_endpoint(instance_name, table_name), row_json,
        "Content-Type" => "application/json")
      case response
      when Net::HTTPSuccess then
        JSON.parse(response.body)
      else
        raise_exception(response.body)
      end
    end

    private

    # BEGIN endpoint definitions
    def instances_endpoint
      "/#{@version}/instances"
    end

    def instance_endpoint(instance_name)
      "#{instances_endpoint}/#{instance_name}"
    end

    def tables_endpoint(instance_name)
      "#{instance_endpoint(instance_name)}/tables"
    end

    def table_endpoint(instance_name, table_name)
      "#{tables_endpoint(instance_name)}/#{table_name}"
    end

    def rows_endpoint(instance_name, table_name)
      "#{table_endpoint(instance_name, table_name)}/rows"
    end

    # END endpoint definitions

    # Generic method that will raise a FijiRestClientError that contains a JSON object
    # with the error in question. In case the exception from the server is not JSON encoded,
    # return a generic JSON message with the message sent by the server so that the client
    # has a consistent interface.
    #
    # param: response_body is the contents of the response.
    def raise_exception(response_body)
      begin
        raise FijiRestClientError.new(JSON.parse(response_body))
      rescue
        # If the exception doesn't parse to JSON properly, then return this generic exception
        json_error= {}
        json_error["exception"] =  response_body
        json_error["status"] =  500
        json_error["exceptionName"]= "internal_error"

        raise FijiRestClientError.new(json_error)
      end
    end

    # Returns a parsed JSON object for the given endpoint and query parameters
    # param: endpoint is the HTTP endpoint
    # param: query_params are query parameters to pass along to the endpoint
    def get_json(endpoint, query_params= {})
      url_query_params =  query_params.map {|k, v| "#{k}=#{CGI.escape(v.to_s)}"}.join("&")
      uri =  URI(@base_uri)
      http =  Net::HTTP.new(uri.host, uri.port)
      response =  http.get("#{endpoint}?#{url_query_params}")
      result =  nil
      case response
      when Net::HTTPSuccess then
        JSON.parse(response.body)
      else
        raise_exception(response.body)
      end
    end
  end
end
